src/toctreeview.cpp

```diff
@@ -0,0 +1,182 @@
+#include "toctreeview.h"
+#include "latexstructure.h"
+#include "latexdocument.h"
+#include "configmanager.h"
+#include "utilsUI.h"
+#include "utilsSystem.h"
+#include "editors.h"
+#include "latexeditorview.h"
+#include "tocitemtree.h"
+#include "tocitemdata.h"
+
+TocTreeView::TocTreeView(const ConfigManager& config, LatexDocuments* docs, QWidget* parent) :
+	QTreeView(parent),
+	documents(docs)
+{
+	header()->hide();
+	setObjectName("TocTree");
+	UtilsUi::enableTouchScrolling(this);
+
+	if (config.indentationInStructure > 0)
+		setIndentation(config.indentationInStructure);
+
+	updateTOC();
+}
+
+void TocTreeView::updateTOC()
+{
+	closedDocs.clear();
+	updateData();
+	expandAll();
+}
+
+QList<TocItem> parseStruct(QString file, StructureEntry* str) {
+	QList<TocItem> result;
+	for (int i = 0; i < str->children.count(); i++) {
+		if (str->children[i]->type != StructureEntry::SE_SECTION && str->children[i]->type != StructureEntry::SE_INCLUDE) continue;
+		TocItem item;
+		item.file = file;
+		item.title = str->children[i]->title;
+		item.level = str->children[i]->level;
+		item.line = str->children[i]->getRealLineNumber();
+		if (str->children[i]->type == StructureEntry::SE_INCLUDE) {
+			if (!item.title.endsWith(".tex", Qt::CaseInsensitive)) item.title.append(".tex");
+			item.title = QFileInfo(file).dir().filePath(item.title);
+			item.level = -1;
+		}
+		if (!result.contains(item)) result.push_back(item);
+		if (str->children[i]->type == StructureEntry::SE_SECTION) result.append(parseStruct(file, str->children[i]));
+	}
+	return result;
+}
+
+struct QDataComparer
+{
+	bool operator()(const QList<QVariant>& a, const QList<QVariant>& b) const
+	{
+		return a[1].toInt() < b[1].toInt();
+	}
+};
+
+QList<TocItem> parseFile(QString filePath)
+{
+	QList<TocItem> result;
+	QFile file(filePath);
+	QString text;
+	if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+		QTextStream stream(&file);
+		stream.setCodec("UTF-8");
+		while (!stream.atEnd())
+			text.append(stream.readLine() + "\n");
+	}
+	file.close();
+	QList<QString> sf; sf << "input"; sf << "section"; sf << "subsection"; sf << "subsubsection";
+	QList<QList<QVariant>> il;
+	for (int i = 0; i < sf.count(); i++) {
+		int pos = text.indexOf("\\" + sf[i]);
+		while (pos >= 0) {
+			QList<QVariant> item;
```

Comment:
this is very unreadable code
using a variant as a replacement for a struct (text,lineNr) which inhibit static code checks.